compiler-plugin/src/main/java/io/ballerina/lib/np/compilerplugin/ConstantExpressionVisitor.java

```diff
@@ -0,0 +1,375 @@
+                "contains constant expressions. (found: '%s')", name));
+        this.constantExpressionDiagnostics.add(diagnostic);
+    }
+
+    @Override
+    public void visit(AnnotAccessExpressionNode annotAccessExpressionNode) {
+        addConstantReferencesDiagnostics(annotAccessExpressionNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(ArrayTypeDescriptorNode arrayTypeDescriptorNode) {
+        addConstantReferencesDiagnostics(arrayTypeDescriptorNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(BasicLiteralNode basicLiteralNode) {
+        // Allowed in constant expressions.
+    }
+
+    @Override
+    public void visit(BuiltinSimpleNameReferenceNode builtinSimpleNameReferenceNode) {
+        // Allowed in constant expressions.
+    }
+
+    @Override
+    public void visit(BinaryExpressionNode binaryExpressionNode) {
+        visitSyntaxNode(binaryExpressionNode.lhsExpr());
+        visitSyntaxNode(binaryExpressionNode.rhsExpr());
+    }
+
+    @Override
+    public void visit(BracedExpressionNode bracedExpressionNode) {
+        visitSyntaxNode(bracedExpressionNode.expression());
+    }
+
+    @Override
+    public void visit(ByteArrayLiteralNode byteArrayLiteralNode) {
+        byteArrayLiteralNode.content().ifPresent(content -> visit(content));
+    }
+
+    @Override
+    public void visit(CheckExpressionNode checkExpressionNode) {
+        addConstantReferencesDiagnostics(checkExpressionNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(ConditionalExpressionNode conditionalExpressionNode) {
+        visitSyntaxNode(conditionalExpressionNode.lhsExpression());
+        visitSyntaxNode(conditionalExpressionNode.middleExpression());
+        visitSyntaxNode(conditionalExpressionNode.endExpression());
+    }
+
+    @Override
+    public void visit(DistinctTypeDescriptorNode distinctTypeDescriptorNode) {
+        addConstantReferencesDiagnostics(distinctTypeDescriptorNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(ErrorConstructorExpressionNode errorConstructorExpressionNode) {
+        addConstantReferencesDiagnostics(errorConstructorExpressionNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(ExplicitAnonymousFunctionExpressionNode explicitAnonymousFunctionExpressionNode) {
+        addConstantReferencesDiagnostics(explicitAnonymousFunctionExpressionNode.functionKeyword().toSourceCode()
+                + explicitAnonymousFunctionExpressionNode.functionSignature().toSourceCode()
+                + explicitAnonymousFunctionExpressionNode.functionBody().toSourceCode());
+    }
+
+    @Override
+    public void visit(ExplicitNewExpressionNode explicitNewExpressionNode) {
+       addConstantReferencesDiagnostics(explicitNewExpressionNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(FieldAccessExpressionNode fieldAccessExpressionNode) {
+        visitSyntaxNode(fieldAccessExpressionNode.expression());
```

Comment:
I intentionally allow this to support field access with constant values.
cc @MaryamZi